### PR TITLE
add rnn-t loss and api

### PR DESCRIPTION
**Pull request by zh794390558**
### PR types

New features
### PR changes

APIs
### Describe

* warprnnt/warprnnt_grad op/kernel
* nn.RNNTLoss / nn.functional.rnnt_loss
* support cpu/gpu/rocm
* c++ and python unittest
* precistion
image

* time cost
image

image

image

paddle 比 warp_transducer 速度快1x以上，可能主要是op实现方式的差异导致的: warp_transducer 采用的python接口实现。

### CODE REVIEW

File: paddle/fluid/operators/unity_build_rule.cmake [hunk 1]
```diff
@@ -354,7 +354,8 @@ register_unity_group(
   recurrent_op.cc
   run_program_op.cc
   softmax_with_cross_entropy_op.cc
-  warpctc_op.cc)
+  warpctc_op.cc
+  warprnnt_op.cc)
```
Comment (chenwhql) [inline paddle/fluid/operators/unity_build_rule.cmake]: 这里应该不需要改动，没有这个文件

File: paddle/phi/kernels/funcs/warp_transducer/include/detail/gpu_rnnt_kernel.h [hunk 2]
```diff
@@ -0,0 +1,388 @@
+// Copyright 2018-2019, Mingkun Huang
+// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (chenwhql) [inline paddle/phi/kernels/funcs/warp_transducer/include/detail/gpu_rnnt_kernel.h]: license写两个是因为这里的实现有咱们自己改的吗？

File: paddle/phi/kernels/funcs/warp_transducer/include/detail/hostdevice.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#if (defined(__HIPCC__) || defined(__CUDACC__))
```
Comment (chenwhql) [inline paddle/phi/kernels/funcs/warp_transducer/include/detail/hostdevice.h]: 这个框架里有吧，是不是可以直接include已有头文件，core/hostdevice.h

File: paddle/phi/kernels/funcs/warp_transducer/include/detail/gpu_rnnt_kernel.h (hunk 2 above)
Comment (zh794390558) [reply paddle/phi/kernels/funcs/warp_transducer/include/detail/gpu_rnnt_kernel.h]: 有对rocm的支持。

File: paddle/phi/kernels/funcs/warp_transducer/include/detail/hostdevice.h (hunk 3 above)
Comment (zh794390558) [reply paddle/phi/kernels/funcs/warp_transducer/include/detail/hostdevice.h]: done

File: paddle/fluid/operators/unity_build_rule.cmake (hunk 1 above)
Comment (zh794390558) [reply paddle/fluid/operators/unity_build_rule.cmake]: ok

File: paddle/phi/kernels/funcs/warp_transducer/include/detail/gpu_rnnt_kernel.h (hunk 2 above)
Comment (zh794390558) [reply paddle/phi/kernels/funcs/warp_transducer/include/detail/gpu_rnnt_kernel.h]: 这个删了吧。只新增了头文件#include "type_def.h"

**Comment by chenwhql** [comment]
> ### PR types
> New features
> 
> ### PR changes
> APIs
> 
> ### Describe
> * warprnnt/warprnnt_grad op/kernel
> * nn.RNNTLoss / nn.functional.rnnt_loss
> * support cpu/gpu/rocm
> * c++ and python unittest
> * precistion
> 
> image
> 
> * time cost
> 
> image
> 
> image
> 
> image
> 
> paddle 比 warp_transducer 速度快1x以上，可能主要是op实现方式的差异导致的: warp_transducer 采用的python接口实现。

目前ops.yaml和python api参数不一致，这个原则上是不行的，具体有两个疑问：
1. num_threads仅在C++端有这个参数，是否必要？我看python端值写死了是1？
2. reduction参数能否也放到C++端，可以复用C++端kernel实现同样的判断逻辑？

**Comment by zh794390558** [comment]
> 1. num_threads仅在C++端有这个参数，是否必要？我看python端值写死了是1？

可以删掉。

**Review by chenwhql — DISMISSED** [review]

**Review by chenwhql — APPROVED** [review]
